client/session.go

```diff
@@ -1625,74 +1661,163 @@ func (s *session) FetchBlocksFromPeers(
+			}
+			errLock.RUnlock()
+			return result
+		}
+		abortError    error
+		setAbortError = func(err error) {
+			errLock.Lock()
+			abortError = err
+			errLock.Unlock()
+		}
+		getAbortError = func() error {
+			errLock.RLock()
+			result := abortError
+			errLock.RUnlock()
+			return result
+		}
+		pending   = int64(len(peers.peers))
+		majority  = atomic.LoadInt32(&s.majority)
+		enqueued  = int32(len(peers.peers))
+		responded int32
+		success   int32
 	)
+	if peers.selfExcluded {
+		state, err := peers.selfHostShardSet.ShardSet().LookupStateByID(shardID)
+		if err == nil && state == shard.Available {
+			// If we excluded ourselves from fetching, we basically treat ourselves
+			// as a successful peer response since we can bootstrap from ourselves
+			// just fine
+			enqueued++
+			success++
+		}
+	}
 
-	pending = int64(len(peers))
 	progress.metadataFetches.Update(float64(pending))
-	for _, peer := range peers {
+	for idx, peer := range peers.peers {
+		idx := idx
 		peer := peer
 
 		wg.Add(1)
 		go func() {
-			defer wg.Done()
-			var err error
-
-			switch version {
-			case FetchBlocksMetadataEndpointV1:
-				err = s.streamBlocksMetadataFromPeer(namespace, shard,
-					peer, start, end, metadataCh, progress)
-			case FetchBlocksMetadataEndpointV2:
-				err = s.streamBlocksMetadataFromPeerV2(namespace, shard,
-					peer, start, end, metadataCh, resultOpts, progress)
-			// Should never happen - we validate the version before this function is
-			// ever called
-			default:
-				err = errInvalidFetchBlocksMetadataVersion
+			defer func() {
+				// Success or error counts towards a response
+				atomic.AddInt32(&responded, 1)
+
+				// Decrement pending
+				progress.metadataFetches.Update(float64(atomic.AddInt64(&pending, -1)))
+
+				// Mark done
+				wg.Done()
+			}()
+
+			var currPageToken pageToken
+			condition := func() bool {
+				level := consistencyLevel
+				majority := int(majority)
+				enqueued := int(enqueued)
+				success := int(atomic.LoadInt32(&success))
+				return !s.readConsistencyAchieved(level, majority, enqueued, success) &&
+					getAbortError() == nil
 			}
+			for condition() {
+				var err error
+				switch version {
+				case FetchBlocksMetadataEndpointV1:
+					currPageToken, err = s.streamBlocksMetadataFromPeer(namespace, shardID,
+						peer, start, end, currPageToken, metadataCh, progress)
+				case FetchBlocksMetadataEndpointV2:
+					currPageToken, err = s.streamBlocksMetadataFromPeerV2(namespace, shardID,
+						peer, start, end, currPageToken, metadataCh, resultOpts, progress)
+				default:
+					// Should never happen - we validate the version before this function is
+					// ever called
+					err = xerrors.NewNonRetryableError(errInvalidFetchBlocksMetadataVersion)
+					setError(idx, err)
+					return
+				}
 
-			if err != nil {
-				errLock.Lock()
-				defer errLock.Unlock()
-				errLen++
-				multiErr = multiErr.Add(err)
+				// Set error or success if err is nil
```

Comment:
This comment is a little confusing because it makes it seem like setError will increment the atomic success variable. I think it would be clearer if you left the comment out or just did

```golang
if err != nil {
  setError(idx, err)
}
```

then there is no ambiguity even if it behaves exactly the same way

Reply:
It actually needs to nil if it's no longer error for this replica so we don't report an old error when we call `getErrors()`.

Reply:
ah ok fair enough. That might be worth a one line comment